dask/dataframe/core.py

```diff
@@ -4408,3 +4369,42 @@ def _repr_data_series(s, index):
     else:
         dtype = str(s.dtype)
     return pd.Series([dtype] + ['...'] * npartitions, index=index, name=s.name)
+
+
+parallel_types = [
+    (pd.Series, Series),
+    (pd.DataFrame, DataFrame),
+    (pd.Index, Index),
+]
+
+
+def _get_return_type(meta):
+    if isinstance(meta, _Frame):
+        meta = meta._meta
+
+    for pd_type, dd_type in reversed(parallel_types):
+        if isinstance(meta, pd_type):
+            return dd_type
```

Comment:
Why not use `Dispatch` here? This would handle type hierarchy's in a robust way (more specific type first), and avoids having to iterate through a list every time.

```python
def _get_return_type(meta):
    if isinstance(meta, _Frame):
        meta = meta._meta

    return get_parallel_type(meta)

get_parallel_type = Dispatch('get_parallel_type')
get_parallel_type.register(object, Scalar)
get_parallel_type.register(pd.Series, Series)
get_parallel_type.register(pd.DataFrame, DataFrame)
get_parallel_type.register(pd.Index, Index)
```

Reply:
Using Dispatch seems more complex than a list.  

I agree though that the type hierarchy problem is potentially an issue. Currently there aren't any subclass relationships (though with something like geopandas, there could be).

Iterating through a list is effectively free here.  Iterating through lists is relatively fast and we only call this when we create new parallel collections.

I think that I'll counter-propose that we go the other way for now and use a `dict` and that, if a subclass relationship shows up in the future then we add code to run a toposort each time.  I think that this is simplest today on all fronts (though a bit risky) and remains simplest on downstream users in the future if we add a toposort later on.

Reply:
I agree that using `Dispatch` isn't needed for the current subclasses, but it's sufficient for the problem, fast, and presents a common extension point (other parts of the extension interface equally sue `Dispatch`). My concern here is mainly for uniformity - no reason to create a new extension mechanism when one already exists.

Reply:
You're right.  It was nicer to just use Dispatch.  I have a slightly different implementation though in that I'm returning the parallel type rather than registering it.